Add  json format output to instance.list

ArangoGutierrez opened the pull request:
**Description of the Pull Request (PR):**

Add  json format output to instance.list

Example output:

```bash
$ singularity instance.list --json
{"instances":[{"instance":web1,"pid":10925,"img":/var/lib/singularity/img/nginx.simg},{"instance":web2,"pid":10963,"img":/var/lib/singularity/img/nginx.simg},{"instance":web3,"pid":15577,"img":/var/lib/singularity/img/nginx.simg}]}
```

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [x] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularityware-admin

ArangoGutierrez committed:
Add  json format output to instance.list

add a json format output to instance.list
```bash
jsonFmt (){
  INSTANCE='"instance"'
  INSTANCE_PID='"pid"'
  INSTANCE_IMG='"img"'
  JSON=""
  JSON_INIT='{"instances":['
  JSON_END=']}'
  COUNT=0


  if [ -z "${1:-}" ]; then
      DAEMONS=`singularity_daemon_glob '*' | sort | uniq`
  else
      DAEMONS=`singularity_daemon_glob "$@" | sort | uniq`
  fi

DLEN=0
for i in $DAEMONS; do
	let "DLEN++"
done
let "DLEN--"

  for i in $DAEMONS; do
      if [ -f "$i" ]; then
          if [ ! -s "$i" ]; then
              rm "$i"
              continue
          fi
          FILE_NAME=`basename "$i"`
          DAEMON_NAME=${FILE_NAME##*-}
          . "$i"
          if [ -n "${DAEMON_PID:-}" ]; then
              PROC_CMDLINE="/proc/${DAEMON_PID:-}/cmdline"
              if [ -e "$PROC_CMDLINE" ]; then
                  PROGNAME=$(<"${PROC_CMDLINE}" tr -d \\0)
                  if [ "${PROGNAME}" != "singularity-instance: $USER [$DAEMON_NAME]" ]; then
                      message WARN "Removing stale daemon file: $i\n"
                      rm -f "$i"
                      continue
                  fi
              else
                  rm -f "$i"
                  continue
              fi
          fi

	if [ "$COUNT" -lt "$DLEN" ]; then
		JSON+="{"${INSTANCE}":"${FILE_NAME}","${INSTANCE_PID}":"${DAEMON_PID}","${INSTANCE_IMG}":"${DAEMON_IMAGE:-}"},"
	else
		JSON+="{"${INSTANCE}":"${FILE_NAME}","${INSTANCE_PID}":"${DAEMON_PID}","${INSTANCE_IMG}":"${DAEMON_IMAGE:-}"}"
	fi

          let "COUNT++"
      fi
  done

  if [ "$COUNT" == 0 ]; then
      exit 1
  fi
  echo "${JSON_INIT}${JSON}${JSON_END}"
}
```

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -49,6 +110,10 @@ while true; do
             fi
         ;;
+        --json)
+          jsonFmt
```
I would like to see this controlled by an environment variable instead of calling the function directly here.  This way you can combine your function with the code below and control whether it prints or constructs a json via the env var.

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -33,6 +33,67 @@ if ! USERID=`id -ru`; then
+DLEN=0
+for i in $DAEMONS; do
+	let "DLEN++"
+done
```
I think you could do this in a one-liner if you echo $DAEMONS with quotes to preserve newlines and use wc.  But I'm not convinced this variable is necessary at all.  I think you could change your if statement in the for loop below to look at the one-liner instead of using the variable.  Might be a slight bit less efficient but would increase readability imho.

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -33,6 +33,67 @@ if ! USERID=`id -ru`; then
+                  continue
+              fi
+          fi
+
```
All of this code is basically just copied and pasted from the code below with the one change of constructing a json instead of printing each line.  That is not going to be very sustainable if we have to change this algorithm.  Can you just wrap the if statement below in another if statement so that it prints if the json environmental variable is not set and it creates the json if the json env var is set?

GodloveD reviewed (changes requested):
Hi @ArangoGutierrez!  Thanks for identifying this new potential feature and for taking the time to implement it!  

I like the idea, but I'd like to see some changes in it's implementation.  I'd really like if you could do this without duplicating any of the existing code.  I think that if you set an environment variable in the switch statement that parses the input options you could could add a single if statement that would control whether the code prints daemon names line by line or constructs json output.  

I'm also going to as @vsoch to look this over and tell me if she thinks the json format is optimal. 

Thanks!

ArangoGutierrez committed:
Changes requested on #1191

Changes requested on #1191

ArangoGutierrez committed:
Update CHANGELOG.md

ArangoGutierrez committed:
"" quotes added 

"" quotes added

ArangoGutierrez committed:
fix unbound variable error

Doing 
```bash
JSONFMT="false"
```
at line 36, error :
```bash
/usr/local/libexec/singularity/cli/instance.list.exec: line 76: JSONFMT: unbound variable
```
got fixed

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -33,6 +33,7 @@ if ! USERID=`id -ru`; then
     exit 255
 fi
 
+JSONFMT="false"
```
Instead of doing this we tend to do something like: 
```
"${JSONFMT:-}"
```
That way a user can set the env var and in their environment and control the format that way rather than passing the argument.

GodloveD replied:
Also a good idea to prepend variables like this with `SINGULARITY_`.  After all, the user might want set this env var and if you export it, you want to make sure that it ends up in the registry for use in the c portion of the code.

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -49,6 +50,10 @@ while true; do
             fi
         ;;
+        --json)
+        JSONFMT="true"
```
Can you indent this please. 😄

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -66,7 +71,14 @@ else
+DLEN=`echo "$DAEMONS"|wc -l`
+let DLEN--
+
+if [[ "${JSONFMT}" == "true" ]]; then
```
In keeping with the conventions we use throughout the code, it would be good to change this to something like:
```
if [ -z "${SINGULARITY_JSONFMT:-}" ]; then
```
I think in most of our code, any value of an env var is taken as true.  It is only false if it is empty or not set.

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -92,12 +104,24 @@ for i in $DAEMONS; do
 
-        printf "%-16s %-8s %s\n" "${FILE_NAME}" "${DAEMON_PID}" "${DAEMON_IMAGE:-}"
+        if [[ "${JSONFMT}" == "true" ]]; then
+          if [ "$COUNT" -lt "$DLEN" ]; then
```
I think we're using 4 indentations throughout right?

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -92,12 +104,24 @@ for i in $DAEMONS; do
 done
 
+if [[ "${JSONFMT}" == "true" ]]; then
+printf "\n\t]\n}\n"
```
I can haz indent plz? 🐱

GodloveD commented on libexec/cli/instance.list.exec:
```diff
@@ -92,12 +104,24 @@ for i in $DAEMONS; do
-        printf "%-16s %-8s %s\n" "${FILE_NAME}" "${DAEMON_PID}" "${DAEMON_IMAGE:-}"
+        if [[ "${JSONFMT}" == "true" ]]; then
+          if [ "$COUNT" -lt "$DLEN" ]; then
+            printf '\t{\n\t\t"instance":"%s",\n\t\t"pid":"%s",\n\t\t"img":"%s"\n\t},\n' "${FILE_NAME}" "${DAEMON_PID}" "${DAEMON_IMAGE:-}"
```
Would it be useful to save this big gnarly string in a variable so that you don't have to repeat it on 2 lines?

GodloveD reviewed (changes requested):
Hey @ArangoGutierrez.  This looks a lot better but I still think it needs a little more love. 

Also can you add the `--json` argument to the `instance.list` subcommand in [etc/bash_completion.d/singularity](etc/bash/completion.d/singularity)

ArangoGutierrez committed:
requested changes on PR #1191

moved 
```bash
{SINGULARITY_JSONFMT} == "true"
```
to 
```bash
-z {SINGULARITY_JSONFMT:-}
```
Also created a var to store the print format for the json
```bash
JSON_PRINTF='\t{\n\t\t"instance":"%s",\n\t\t"pid":"%s",\n\t\t"img":"%s"\n\t}'
```

ArangoGutierrez committed:
Added the json flag to bash-completion

Added the json flag to bash-completion

ArangoGutierrez committed:
Update instance.list.exec

GodloveD committed:
added help for instance.list --json, added short (-j) option, changed env var to be consistent with inspect, and changed 2 space indents to 4 space indents

GodloveD reviewed (approved):
Hey @ArangoGutierrez, I took the liberty of making a few more changes to this myself.  I'm happy with this now.  If somebody else can take a look I think we're good to merge.  @carlm1234 do you want to give it a once over?
